Ayush sprint1

AyushChoudhury opened the pull request:
I merged my shader loading changes with the master branch. There are still a few issues like if the object does not have its own color, the ads will not have any ambient lighting. And there is no way to add a shader and define new uniform variables for it. But there are no code-breaking bugs that I found. I have not added tests for loading shaders yet.

AyushChoudhury committed:
Loads Shader file

Shaders are now gotten from a file, and user can select a new vertex and fragment shader to use.

AyushChoudhury committed:
Shaders

The ADS shaders can now be loaded. They are test.vert and test.frag.

AyushChoudhury committed:
shader file rename

renamed the lighting shaders from test.vert and test.frag to ads.vert and ads.frag

AyushChoudhury committed:
Merge branch 'main' into Ayush-Sprint1

tigerman9854 commented on ModelViewer/ModelViewer/ModelViewer.vcxproj:
```diff
@@ -74,7 +74,7 @@
     </PostBuildEvent>
   </ItemDefinitionGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'" Label="QtSettings">
-    <QtInstall>msvc2019_64</QtInstall>
```
Let's not commit this change.  Everyone else has their Qt named "msvc2019_64"

AyushChoudhury replied:
yup, that was a mistake

tigerman9854 commented on ModelViewer/ModelViewer/ViewerGraphicsWindow.h:
```diff
@@ -16,13 +16,19 @@ class ViewerGraphicsWindow : public OpenGLWindow
     float viewportX = 0;
     float viewportY = 0;
 
+    QMatrix4x4 modelview;
```
I'm not a fan of the duplicated code related to the modelView matrix, but it's okay for now.  Once Jonathan reworks the mouse controls, we will have separate rotation, translation, and scale matrices which can be used to compute the modelView when needed.

AyushChoudhury replied:
agreed. I just wasn't sure how to go from modelviewprojection matrix back to modelview matrix

tigerman9854 commented on ModelViewer/ModelViewerRun/ModelViewerRun.vcxproj:
```diff
@@ -66,7 +66,7 @@
     </Link>
   </ItemDefinitionGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'" Label="QtSettings">
-    <QtInstall>msvc2019_64</QtInstall>
```
Same comment as above, let's not commit this change because everyone else has their Qt named "msvc2019_64"

tigerman9854 commented on ModelViewer/ModelViewer/ViewerGraphicsWindow.cpp:
```diff
@@ -57,6 +57,136 @@ bool ViewerGraphicsWindow::loadModel(QString filepath) {
+        QString error = m_program->log();
+        qDebug() << error << endl;
+        currentVertFile = vertfilepath;
+        //m_program->addShaderFromSourceFile(QOpenGLShader::Vertex, currentVertFile);
```
I think we delete this code since it's commented out, right?

AyushChoudhury replied:
yeah, feel free to delete anything that's commented out

tigerman9854 commented on ModelViewer/ModelViewer/ViewerGraphicsWindow.cpp:
```diff
@@ -57,6 +57,136 @@ bool ViewerGraphicsWindow::loadModel(QString filepath) {
+    m_uSpecularColor = m_program->uniformLocation("uSpecularColor");
+    m_uShininess = m_program->uniformLocation("uShininess");
+    /*
+    m_posAttr = newProgram->attributeLocation("posAttr");
```
Same as above, can this just be deleted?

AyushChoudhury replied:
For some reason the code only works when I reassign the uniform variables every time we load a shader. I will look into this, but for now removing it would probably break the code.

tigerman9854 replied:
I just meant the commented out parts.

tigerman9854 commented on ModelViewer/ModelViewer/ViewerGraphicsWindow.cpp:
```diff
@@ -57,6 +57,136 @@ bool ViewerGraphicsWindow::loadModel(QString filepath) {
+        return false;
+    }
+    if (vertfilepath.isEmpty()) {
+        vertfilepath = QFileDialog::getOpenFileName(nullptr, "Load Vertex Shader", "../Data/Shaders/", "");
```
If you do `QFileDialog::getOpenFileName(nullptr, "Load Vertex Shader", "../Data/Shaders/", "*.vert");`, it will only show .vert files.  This might be nice.

AyushChoudhury replied:
yeah, I wasn't sure whether we wanted to support other text file formats or not. But that would be fine

tigerman9854 commented on ModelViewer/ModelViewer/ViewerGraphicsWindow.cpp:
```diff
@@ -57,6 +57,136 @@ bool ViewerGraphicsWindow::loadModel(QString filepath) {
+    //QOpenGLShaderProgram *newProgram = new QOpenGLShaderProgram(this);
+    m_program->removeAllShaders();
+
+    m_program->addShaderFromSourceFile(QOpenGLShader::Vertex, vertfilepath);
```
Wherever we call addShaderFromSourceFile(), we need to check the return value.  The return value indicates if compilation was successful or not.

AyushChoudhury replied:
Okay. I just decided to stop checking for compilation errors, because loading in one shader at a time breaks you code more ofte than not, and we needed to remember the shader even if it breaks your code because it might work once you load another shader.

tigerman9854 commented on ModelViewer/Data/Shaders/ads.frag:
```diff
@@ -0,0 +1,36 @@
+varying lowp vec4 col;
```
In order to compile on NVIDIA, the ADS shaders need to have #version 130 at the top.  This allows use of in/out keywords

AyushChoudhury replied:
oh, okay. I'll make this change

AyushChoudhury committed:
Pull request recommended changes

removed comments, check for errors whenever loading a shader file (other than ones that have already been loaded), fixed qt install version name, added #version 130 to shader files.
